### PR TITLE
[WIP] Refactoring : RC1 Release

### PR DESCRIPTION
**Pull request by ikwattro**
This branch is a big refactoring, dropping code that was just impossible to maintain and to read. Re-starting repository and hydration from scratch and following some Doctrine flows.

NB: for now, all entities will be proxied.

This branch / PR marked as done will trigger an RC1 release.

### CODE REVIEW

File: src/Hydration/ObjectHydration.php [hunk 1]
```diff
@@ -57,11 +50,20 @@ class ObjectHydration
     private $entityManager;
 
     /**
+     * @var string
+     */
+    private $className;
+
```
Comment (Nyholm) [inline src/Hydration/ObjectHydration.php]: Add comment to explain what data this is.

File: src/Hydration/ObjectHydration.php [hunk 2]
```diff
@@ -57,11 +50,20 @@ class ObjectHydration
...
+    private $classMetadata;
+
+    /**
      *
      * @param EntityManager $entityManager
```
Comment (Nyholm) [inline src/Hydration/ObjectHydration.php]: Update doc block

File: src/Hydration/ObjectHydration.php [hunk 3]
```diff
@@ -379,6 +386,8 @@ private function hydrateNode(Node $node, $className = null, $refresh = false)
             $instance = $cm->newInstance();
         }
 
+        $instance = $this->entityManager->getProxyFactory($this->classMetadata)->fromNode($node);
```
Comment (Nyholm) [inline src/Hydration/ObjectHydration.php]: We should remove previous assignments to $instance.

File: src/Hydrator/EntityHydrator.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+use GraphAware\Neo4j\OGM\Common\Collection;
+use GraphAware\Neo4j\OGM\EntityManager;
+use GraphAware\Neo4j\OGM\Metadata\NodeEntityMetadata;
+
+class EntityHydrator
+{
+    /**
+     * @var EntityManager
+     */
+    protected $_em;
```
Comment (Nyholm) [inline src/Hydrator/EntityHydrator.php]: Should be private and one should remove the underscore. Same with $_classMetadata

File: src/Hydrator/EntityHydrator.php [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        foreach ($dbResult->records() as $record) {
+            $this->hydrateRecord($record, $result);
+        }
+
+        return $result;
+    }
+
+    /**
+     * @param Result $dbResult
```
Comment (Nyholm) [inline src/Hydrator/EntityHydrator.php]: Update doc block

File: src/Metadata/RelationshipMetadata.php [hunk 6]
```diff
@@ -58,13 +63,14 @@
      * @param bool                                           $isLazy
      * @param OrderBy                                        $orderBy
```
Comment (Nyholm) [inline src/Metadata/RelationshipMetadata.php]: Update doc block

File: src/Persisters/BasicEntityPersister.php [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+namespace GraphAware\Neo4j\OGM\Persisters;
+
+use GraphAware\Common\Cypher\Statement;
+use GraphAware\Neo4j\OGM\EntityManager;
+use GraphAware\Neo4j\OGM\Metadata\NodeEntityMetadata;
+use GraphAware\Neo4j\OGM\Util\DirectionUtils;
+
+class BasicEntityPersister
+{
+    protected $_className;
```
Comment (Nyholm) [inline src/Persisters/BasicEntityPersister.php]: Same here, Use private and remove underscore.

File: src/UnitOfWork.php [hunk 8]
```diff
@@ -438,6 +442,7 @@ private function checkRelationshipEntityDeletions($entity)
     public function detectRelationshipReferenceChanges()
     {
         foreach ($this->managedRelationshipReferences as $oid => $reference) {
+            //var_dump($oid);
```
Comment (Nyholm) [inline src/UnitOfWork.php]: Remove debug code

File: src/Hydration/ObjectHydration.php (hunk 1 above)
Comment (ikwattro) [inline src/Hydration/ObjectHydration.php]: this class will not exist anymore

File: src/Hydration/ObjectHydration.php (hunk 2 above)
Comment (ikwattro) [inline src/Hydration/ObjectHydration.php]: same here

File: src/Hydration/ObjectHydration.php (hunk 3 above)
Comment (ikwattro) [inline src/Hydration/ObjectHydration.php]: this class is not used anymore

File: src/Hydrator/EntityHydrator.php (hunk 4 above)
Comment (ikwattro) [inline src/Hydrator/EntityHydrator.php]: what is the rationale for the underscore ?

File: src/Hydrator/EntityHydrator.php (hunk 5 above)
Comment (ikwattro) [inline src/Hydrator/EntityHydrator.php]: ?

**Pull request merged**